File: segy_2d_viewer/src/segy_reader.py
"""
SEG-Y 파일 읽기 및 파싱 모듈
"""
import numpy as np
import segyio
from typing import Optional, Tuple


class SegyReader:
    """SEG-Y 파일을 읽고 데이터를 처리하는 클래스"""

    def __init__(self):
        self.filename: Optional[str] = None
        self.segy_file: Optional[segyio.SegyFile] = None
        self.data: Optional[np.ndarray] = None
        self.num_traces: int = 0
        self.num_samples: int = 0
        self.sample_rate: float = 0.0

    def load_file(self, filename: str) -> bool:
        """
        SEG-Y 파일을 로드합니다.

        Args:
            filename: SEG-Y 파일 경로

        Returns:
            성공 여부
        """
        try:
            self.filename = filename
            self.segy_file = segyio.open(filename, ignore_geometry=True)

            # 기본 정보 읽기
            self.num_traces = len(self.segy_file.trace)
            self.num_samples = len(self.segy_file.samples)
            self.sample_rate = segyio.tools.dt(self.segy_file) / 1000.0  # ms to seconds

            # 데이터 읽기
            self.data = np.zeros((self.num_traces, self.num_samples), dtype=np.float32)
            for i, trace in enumerate(self.segy_file.trace):
                self.data[i, :] = trace

            # 데이터 정규화 (시각화를 위해)
            self._normalize_data()

            return True

        except Exception as e:
            print(f"Error loading SEG-Y file: {e}")
            return False

    def _normalize_data(self):
        """데이터를 정규화합니다 (시각화를 위해)"""
        if self.data is not None:
            # 각 트레이스별로 정규화
            for i in range(self.num_traces):
                trace = self.data[i, :]
                max_abs = np.max(np.abs(trace))
                if max_abs > 0:
                    self.data[i, :] = trace / max_abs

    def get_data(self) -> Optional[np.ndarray]:
        """
        정규화된 데이터를 반환합니다.

        Returns:
            데이터 배열 (num_traces x num_samples)
        """
        return self.data

    def get_trace(self, trace_index: int) -> Optional[np.ndarray]:
        """
        특정 트레이스 데이터를 반환합니다.

        Args:
            trace_index: 트레이스 인덱스

        Returns:
            트레이스 데이터
        """
        if self.data is not None and 0 <= trace_index < self.num_traces:
            return self.data[trace_index, :]
        return None

    def get_dimensions(self) -> Tuple[int, int]:
        """
        데이터 차원을 반환합니다.

        Returns:
            (num_traces, num_samples)
        """
        return (self.num_traces, self.num_samples)

    def get_sample_rate(self) -> float:
        """
        샘플링 레이트를 반환합니다.

        Returns:
            샘플링 레이트 (초 단위)
        """
        return self.sample_rate

    def get_time_axis(self) -> np.ndarray:
        """
        시간 축 데이터를 반환합니다.

        Returns:
            시간 배열 (초 단위)
        """
        if self.segy_file is not None:
            return self.segy_file.samples / 1000.0  # ms to seconds
        return np.array([])

    def close(self):
        """파일을 닫습니다."""
        if self.segy_file is not None:
            self.segy_file.close()
            self.segy_file = None

File: segy_2d_viewer_rust/src/segy_reader.rs
// SEG-Y Reader Module - mmap + rayon based
use anyhow::{Context, Result};
use byteorder::{BigEndian, ByteOrder};
use memmap2::Mmap;
use rayon::prelude::*;
use std::fs::File;
use std::path::Path;

pub struct SegyReader {
    pub data: Vec<Vec<f32>>,           // [samples][traces]
    pub raw_data: Vec<Vec<f32>>,       // raw data before normalization
    pub num_traces: usize,
    pub num_samples: usize,
    pub sample_rate: f32,
}

impl SegyReader {
    pub fn new() -> Self {
        Self {
            data: Vec::new(),
            raw_data: Vec::new(),
            num_traces: 0,
            num_samples: 0,
            sample_rate: 0.0,
        }
    }

    pub fn load_file<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
        let file = File::open(&path).context("Failed to open SEG-Y file")?;
        let mmap = unsafe { Mmap::map(&file).context("Failed to mmap file")? };

        // Read binary header
        if mmap.len() < 3600 {
            anyhow::bail!("File too small to be a valid SEG-Y file");
        }

        // Read number of samples (bytes 3220-3221, 2 bytes, big-endian)
        self.num_samples = BigEndian::read_u16(&mmap[3220..3222]) as usize;

        // Read sample interval (bytes 3216-3217, microseconds)
        let sample_interval_us = BigEndian::read_u16(&mmap[3216..3218]) as f32;
        self.sample_rate = sample_interval_us / 1_000_000.0; // to seconds

        // Calculate number of traces
        let header_size = 3600;
        let trace_size = 240 + self.num_samples * 4;
        self.num_traces = (mmap.len() - header_size) / trace_size;

        println!(
            "Loading SEG-Y: {} traces, {} samples",
            self.num_traces, self.num_samples
        );

        // Parallel load traces
        let trace_indices: Vec<usize> = (0..self.num_traces).collect();
        let traces: Vec<Vec<f32>> = trace_indices
            .par_iter()
            .map(|&trace_idx| {
                Self::load_trace(&mmap, trace_idx, self.num_samples, header_size, trace_size)
            })
            .collect();

        // Transpose: Vec<Vec<f32>> where outer is traces -> [samples][traces]
        self.raw_data = vec![vec![0.0; self.num_traces]; self.num_samples];
        for (trace_idx, trace) in traces.iter().enumerate() {
            for (sample_idx, &value) in trace.iter().enumerate() {
                self.raw_data[sample_idx][trace_idx] = value;
            }
        }

        // Normalize data
        self.data = self.normalize_data(&self.raw_data);

        Ok(())
    }

    fn load_trace(
        mmap: &Mmap,
        trace_idx: usize,
        num_samples: usize,
        header_size: usize,
        trace_size: usize,
    ) -> Vec<f32> {
        let mut trace_data = vec![0.0; num_samples];
        let trace_offset = header_size + trace_idx * trace_size + 240; // skip trace header

        for i in 0..num_samples {
            let offset = trace_offset + i * 4;
            if offset + 4 > mmap.len() {
                break;
            }

            let ibm_bytes = &mmap[offset..offset + 4];
            trace_data[i] = Self::ibm_to_ieee(ibm_bytes);
        }

        trace_data
    }

    fn ibm_to_ieee(bytes: &[u8]) -> f32 {
        let ibm_int = BigEndian::read_u32(bytes);

        let sign = (ibm_int >> 31) & 1;
        let exponent = ((ibm_int >> 24) & 0x7f) as i32;
        let mantissa = ibm_int & 0x00ffffff;

        if mantissa == 0 {
            return 0.0;
        }

        // Convert mantissa to [0.0625, 1.0)
        let mut value = mantissa as f64 / 16777216.0; // 2^24

        // Calculate exponent (overflow protection)
        let exp_val = exponent - 64;
        if exp_val > 60 || exp_val < -60 {
            return 0.0;
        }

        value *= 16.0_f64.powi(exp_val);

        if sign == 1 {
            value = -value;
        }

        // Check for inf/nan
        if value.is_finite() {
            value as f32
        } else {
            0.0
        }
    }

    fn normalize_data(&self, data: &[Vec<f32>]) -> Vec<Vec<f32>> {
        let mut normalized = data.to_vec();

        // Normalize each trace (column)
        for trace_idx in 0..self.num_traces {
            let mut max_abs = 0.0f32;

            // Find max absolute value in this trace
            for sample_idx in 0..self.num_samples {
                let val = data[sample_idx][trace_idx].abs();
                if val.is_finite() && val > max_abs {
                    max_abs = val;
                }
            }

            // Normalize
            if max_abs > 0.0 {
                for sample_idx in 0..self.num_samples {
                    let val = data[sample_idx][trace_idx];
                    normalized[sample_idx][trace_idx] = if val.is_finite() {
                        val / max_abs
                    } else {
                        0.0
                    };
                }
            } else {
                for sample_idx in 0..self.num_samples {
                    normalized[sample_idx][trace_idx] = 0.0;
                }
            }
        }

        normalized
    }

    pub fn get_trace(&self, trace_idx: usize) -> Option<Vec<f32>> {
        if trace_idx >= self.num_traces {
            return None;
        }

        let mut trace = vec![0.0; self.num_samples];
        for sample_idx in 0..self.num_samples {
            trace[sample_idx] = self.raw_data[sample_idx][trace_idx];
        }
        Some(trace)
    }
}
